### PR TITLE
[FSDP()] Register root pre-forward hook

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89573
* __->__ #89572

- This PR registers the FSDP root pre-forward hook as a module forward pre-hook following the recently added support for kwargs for those hooks.
- This PR also passes `prepend=True` for the normal (not root) pre-forward hook. This is not strictly required for this PR, but I believe it is needed for composability with activation checkpointing. (We want to run FSDP logic on the outside and AC logic on the inside, just like how we recommend `FSDP(AC(module))` for the wrapper versions.)

Fun fact: I originally chose the `[FSDP()]` prefix in the PR titles when we still referred to composable FSDP as functional-like FSDP, in which case `FSDP()` approximated "functional FSDP". I am preserving this usage to make searching for PRs relating to composable FSDP easier.

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -317,8 +317,8 @@ def _post_forward_reshard(
 def _root_pre_forward(
     state: _FSDPState,
     module: nn.Module,
-    *args,
-    **kwargs,
+    args,
+    kwargs,
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:321]: nit: types? (no need to update this PR)

File: torch/distributed/fsdp/_runtime_utils.py [hunk 2]
```diff
@@ -928,25 +930,27 @@ def _register_post_forward_hooks(
...
-    # TODO (awgu): This requires kwarg support for hooks registered by
-    ``register_forward_pre_hook()``. ``_root_pre_forward()`` does not have the
-    supported hook signature right now.
+    Registers root pre-forward hook on ``module``, which should be the local
+    FSDP root.
+
+    NOTE: For the current composable FSDP design, we have each application of
+    ``fully_shard()`` to a module to indicate that that module is the local
+    FSDP root. We may remove this assumption in the future, in which case we
+    will need to register this root pre-forward hook on any candidate module
```
Comment (mrshenli) [inline torch/distributed/fsdp/_runtime_utils.py:944]: curious, are you referring to after having a dedicated API marking the global root, `fully_shard` might no longer mark the FSDP root? Or is this referring to sth else?

**Review by mrshenli — APPROVED** [review]
Thanks for updating!

File: torch/distributed/fsdp/_runtime_utils.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:944]: For context, let me recap the manual wrapping approach using the `FullyShardedDataParallel` wrapper:
- The user may wrap submodules with `FullyShardedDataParallel`, each following some eager initialization semantics (like constructing and sharding the instance's `FlatParameter`).
- Then, the first forward triggers some lazy initialization semantics (via `_lazy_init()`). Each `FullyShardedDataParallel` instance runs `_lazy_init()` in its pre-forward logic, and the first to call it is marked as the root instance (taking advantage of the module tree structure).

We may imagine a similar approach for `fully_shard()`.
- The user may apply `fully_shard()` to submodules, each following some eager initialization semantics.
- Then, the first forward triggers some lazy initialization semantics (via `_lazy_init()`). We would need to register `_root_pre_forward()` on the module passed to `fully_shard()` _for each_ application of `fully_shard()`. This is what the comment is referring to.

The current design assumes that each application of `fully_shard()` yields a standalone FSDP subtree without any data structure sharing across applications of `fully_shard()`. IIUC, this suffices for H2 2022.

A new design may want to share data structures across applications of `fully_shard()`. If we pursue such a design, we will encounter the case described above that is similar to manual wrapping.

Note that the multiple applications of `fully_shard()` differs from manual wrapping using `FullyShardedDataParallel` in that each application of `fully_shard()` can construct multiple `FlatParameter`s, not just one, since each can represent a logical auto wrapping.

File: torch/distributed/fsdp/_runtime_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:321]: I am actually not exactly sure what the types would be.

For `kwargs`, would it be `Dict[str, Any]`? For `args`, would it be `Iterable[Any]`?

(I am guessing here that an object is an `Iterable` iff it can be unpacked with the `*` operator.)
